--blkio-weight failing in OCI mode only on RHEL8

**Version of Singularity**

3.11.0

**Describe the bug**

On RHEL8, the `--blkio-weight` restriction flag is failing in `--oci` mode action tests only.

```
                --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_root_cgroups (5.07s)
                    --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_root_cgroups/systemd (2.57s)
                        --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_root_cgroups/systemd/OCIRoot (1.38s)
                            --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_root_cgroups/systemd/OCIRoot/blkio-weight (0.13s)
                    --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_root_cgroups/cgroupfs (2.43s)
                        --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_root_cgroups/cgroupfs/OCIRoot (1.34s)
                            --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_root_cgroups/cgroupfs/OCIRoot/blkio-weight (0.12s)
```

The error seen is...

```
    === RUN   TestE2E/SEQ/CGROUPS/action_flags_root_cgroups/systemd/OCIRoot/blkio-weight
        singularitycmd.go:698: Running command "/usr/bin/singularity exec --oci --blkio-weight 50 -B /sys/fs/cgroup oci-archive:/tmp/stest.3502444847/oci.tar /bin/sh -c cat /sys/fs/cgroup/blkio$(cat /proc/self/cgroup | grep '[,:]blkio[,:]' | cut -d ':' -f 3)/blkio.bfq.weight"
        singularitycmd.go:698:
            "/usr/bin/singularity exec --oci --blkio-weight 50 -B /sys/fs/cgroup oci-archive:/tmp/stest.3502444847/oci.tar /bin/sh -c cat /sys/fs/cgroup/blkio$(cat /proc/self/cgroup | grep '[,:]blkio[,:]' | cut -d ':' -f 3)/blkio.bfq.weight" output:
            2023/03/07 12:17:51  info unpack layer: sha256:1487bff95222881565c7c063129c2c2ce3d6fc4d14796ac7627bd1f167bc5621
            opening file `blkio.weight` for writing: No such file or directory
  
        singularitycmd.go:698: got 1 as exit code and was expecting 0: exit status 1
            waiting for command "/usr/bin/singularity exec --oci --blkio-weight 50 -B /sys/fs/cgroup oci-archive:/tmp/stest.3502444847/oci.tar /bin/sh -c cat /sys/fs/cgroup/blkio$(cat /proc/self/cgroup | grep '[,:]blkio[,:]' | cut -d ':' -f 3)/blkio.bfq.weight"
            github.com/sylabs/singularity/e2e/internal/e2e.TestEnv.RunSingularity.func1
                    github.com/sylabs/singularity/e2e/internal/e2e/singularitycmd.go:673
            github.com/sylabs/singularity/e2e/internal/e2e.Privileged.func1
                    github.com/sylabs/singularity/e2e/internal/e2e/priv.go:84
            github.com/sylabs/singularity/e2e/internal/e2e.TestEnv.RunSingularity.func2
                    github.com/sylabs/singularity/e2e/internal/e2e/singularitycmd.go:698
            runtime.goexit
                    runtime/asm_amd64.s:1598
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -31,6 +31,8 @@
   `<library>.so[.version]` files are listed by `ldconfig -p`.
 - Fix systemd cgroup manager error when running a container as a non-root
   user with `--oci`, on systems with cgroups v1 and `runc`.
+- Fix joining cgroup of instance started as root, with cgroups v1,
+  non-default cgroupfs manager, and no device rules.
 
 ### Changed defaults / behaviours
 

--- a/e2e/cgroups/cgroups.go
+++ b/e2e/cgroups/cgroups.go
@@ -426,8 +426,11 @@ var resourceFlagTests = []resourceFlagTest{
 		args:            []string{"--blkio-weight", "50"},
 		expectErrorCode: 0,
 		controllerV1:    "blkio",
-		// This is the new path. Older kernels may have only `blkio.weight`
-		resourceV1:   "blkio.bfq.weight",
+		// Could be `blkio.bfq.weight` if bfq is available. However, under
+		// cgroups v1 older crun will not set blkio.bfq.weight, so only test
+		// with blkio.weight.
+		// Ref: https://github.com/containers/crun/issues/1157
+		resourceV1:   "blkio.weight",
 		expectV1:     "50",
 		delegationV2: "io",
 		resourceV2:   "io.bfq.weight",

--- a/internal/pkg/cgroups/manager_linux.go
+++ b/internal/pkg/cgroups/manager_linux.go
@@ -131,7 +131,7 @@ func (m *Manager) UpdateFromSpec(resources *specs.LinuxResources) (err error) {
 	opts := &lcspecconv.CreateOpts{
 		CgroupName:       m.group,
 		UseSystemdCgroup: false,
-		RootlessCgroups:  false,
+		RootlessCgroups:  os.Getuid() != 0,
 		Spec:             spec,
 	}
 
@@ -140,10 +140,10 @@ func (m *Manager) UpdateFromSpec(resources *specs.LinuxResources) (err error) {
 		return fmt.Errorf("could not create cgroup config: %w", err)
 	}
 
-	// runc/libcontainer/cgroups for v2 defaults to a deny-all policy, while
+	// runc/libcontainer/cgroups defaults to a deny-all policy, while
 	// singularity has always allowed access to devices by default. If no device
 	// rules are provided in the spec, then skip setting them so the deny-all is
-	// not applied.
+	// not applied when we update the cgroup.
 	if len(resources.Devices) == 0 {
 		lcConfig.SkipDevices = true
 	}
@@ -308,12 +308,15 @@ func newManager(resources *specs.LinuxResources, group string, systemd bool) (ma
 		return nil, fmt.Errorf("could not create cgroup config: %w", err)
 	}
 
-	// runc/libcontainer/cgroups for v2 defaults to a deny-all policy, while
-	// singularity has always allowed access to devices by default. If no device
-	// rules are provided in the spec, then skip setting them so the deny-all is
-	// not applied.
+	// runc/libcontainer/cgroups defaults to a deny-all policy, while
+	// singularity has always allowed access to devices by default.
 	if len(resources.Devices) == 0 {
-		lcConfig.SkipDevices = true
+		resources.Devices = []specs.LinuxDeviceCgroup{
+			{
+				Allow:  true,
+				Access: "rwm",
+			},
+		}
 	}
 
 	cgroup, err := lcmanager.New(lcConfig)

